### PR TITLE
[Model] Implemented SubgraphX Explainer for Heterogeneous graph

### PR DESCRIPTION
**Pull request by kunmukh**
## Description
Implemented SubgraphX Explainer (https://arxiv.org/pdf/2102.05152.pdf) for Heterogeneous Graphs. 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md)

## Changes
Added a new class HeteroSubgraphX for Heterogeneous graph explanation.

The output of the model has been manually verified on MUTAG dataset. A visualization script was created to visualize the mutag dataset (external ref: https://gist.github.com/ndbaker1/4eec321173fe1dfaff458b957ebc9a1b). Unfortunately there's no ground truth for the explanation, so we cannot quantify the model's performance. Instead, we manually investigated the results and confirms that it aligns well with human intuition.

Inline documentation is added following the original style, and an example of using the module is included in the documentation.

### CODE REVIEW

**Comment by mufeili** [comment]
@dgl-bot

File: python/dgl/nn/pytorch/explain/subgraphx.py [hunk 1]
```diff
@@ -364,3 +365,447 @@ def explain_graph(self, graph, feat, target_class, **kwargs):
                 best_immediate_reward = best_leaf.immediate_reward
 
         return best_leaf.nodes
+
+
+class HeteroSubgraphX(nn.Module):
+    r"""SubgraphX from `On Explainability of Graph Neural Networks via Subgraph
+    Explorations <https://arxiv.org/abs/2102.05152>`
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/subgraphx.py:372]:
```suggestion
    Explorations <https://arxiv.org/abs/2102.05152>`__, adapted for heterogeneous graphs
```

Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/subgraphx.py]: done, thanks for catching!

**Comment by mufeili** [comment]
@dgl-bot

**Comment by kunmukh** [comment]
@ndbaker1 It looked at the failed logs, and it looks like a simple error to fix. Lets fix this and also added checks to ensure that we don't miss this case.

```
FAILED tests/python/pytorch/nn/test_nn.py::test_heterosubgraphx[2-5-idtype0-g0] - RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!
```

**Comment by mufeili** [comment]
@dgl-bot

**Comment by mufeili** [comment]
@dgl-bot

**Comment by kunmukh** [comment]
@mufeili  I don't think this error is from our end. Would you please take a look?

```shell
04:57:31] /root/jenkins/workspace/dgl_PR-5530/src/rpc/tensorpipe/tp_communicator.cc:37: Failed to connect to receiver[tcp://172.17.0.7:16300].
```

**Comment by mufeili** [comment]
@dgl-bot

**Comment by mufeili** [comment]
@dgl-bot

File: python/dgl/nn/pytorch/explain/subgraphx.py [hunk 2]
```diff
@@ -364,3 +365,453 @@ def explain_graph(self, graph, feat, target_class, **kwargs):
...
+        ...     logits = model(g, feat)
+        ...     loss = F.cross_entropy(logits, th.tensor([1]))
+        ...     optimizer.zero_grad()
+        ...     loss.backward()
+        ...     optimizer.step()
+
+        >>> # Explain for the graph
+        >>> explainer = HeteroSubgraphX(model, num_hops=1)
+        >>> explainer.explain_graph(g, feat, target_class=1)
+        {'game': tensor([0, 1]), 'user': tensor([1, 2])}
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/subgraphx.py:779]: It seems that I got `tensor([1, 2])` for `user`. Is this due to randomness?

File: python/dgl/nn/pytorch/explain/subgraphx.py [hunk 3]
```diff
@@ -364,3 +365,453 @@ def explain_graph(self, graph, feat, target_class, **kwargs):
...
+        ...     def __init__(self, in_dim, num_classes, canonical_etypes):
+        ...         super(Model, self).__init__()
+        ...         self.etype_weights = nn.ModuleDict(
+        ...             {
+        ...                 "_".join(c_etype): nn.Linear(in_dim, num_classes)
+        ...                 for c_etype in canonical_etypes
+        ...             }
+        ...         )
+        ...
+        ...     def forward(self, graph, feat, eweight=None):
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/subgraphx.py:731]: For this example, you don't need the `eweight` argument and related logic, right?

File: python/dgl/nn/pytorch/explain/subgraphx.py [hunk 4]
```diff
@@ -364,3 +365,453 @@ def explain_graph(self, graph, feat, target_class, **kwargs):
...
+
+    def shapley(self, subgraph_nodes):
+        r"""Compute Shapley value with Monte Carlo approximation.
+
+        Parameters
+        ----------
+        subgraph_nodes : dict[str, Tensor]
+            The dictionary that associates tensor node ids (values) with
+            respective node types (keys) present in graph which are associated
+            with this tree node.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/subgraphx.py:447]: subgraph_nodes[nty] gives the tensor node IDs of node type nty in the subgraph, which are associated with this tree node

File: python/dgl/nn/pytorch/explain/subgraphx.py [hunk 5]
```diff
@@ -364,3 +365,453 @@ def explain_graph(self, graph, feat, target_class, **kwargs):
...
+                )
+
+                # Follow steps for backtracking to original graph node ids
+                # 1. retrieve instanced homograph from connected-component homograph
+                # 2. retrieve instanced heterograph from instanced homograph
+                # 3. retrieve hetero-subgraph from instanced heterograph
+                # 4. retrieve orignal graph ids from subgraph node ids
+                cc_nodes = {
+                    ntype: subg.ndata[NID][ntype][
+                        new_subg.ndata[NID][ntype][
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/subgraphx.py:606]: This may not be correct. Use dgl.NTYPE to also backtrack node types.

Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/subgraphx.py:605]: I believe the `dgl.NTYPE` is used already when using `to_heterogeneous`?  
but here is my thought process on the matter as I'm not yet familiar with the internals if anyone has a cleaner solution:

> example scenario is from MUTAG dataset

say there is an initial graph like the following:

```
{
    'C': tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14]),
    'N': tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15]),
    'O': tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17])
}
```

say the current MCTS exploration subgraph looks like so:

```
{
    'C': tensor([ 1,  2,  3,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14]),
    'N': tensor([15]),
    'O': tensor([16, 17])
}
```

the relative mapping shows up in dgl as the following:

```
{
    '_ID': {
        'C': tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12]),
        'N': tensor([0]),
        'O': tensor([0, 1])
    }
}
```

In the process of generating more MCTS candidates, we remove `C[7]` (which is really `C-9` of the original graph) giving us a new subgraph instance:

```
{
    '_ID': {
        'C': tensor([ 0,  1,  2,  3,  4,  5,  6,  8,  9, 10, 11, 12]),
        'N': tensor([0]),
        'O': tensor([0, 1])
    }
}
```

From here we convert into a homograph and compute the largest connected components using networkX. Then we obtain the nods in a heterograph format.

```
# dgl homograph
{
    '_ID': tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11,  0,  0,  1]),
    '_TYPE': tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 2])
}

# networkX, resulting connected component nids:
[7, 8, 9, 10, 11, 12, 13, 14]

# dgl.subgraph using above nids mask
{
    '_ID': tensor([ 7,  8,  9, 10, 11, 12, 13, 14]),
    '_TYPE': tensor([0, 0, 0, 0, 0, 1, 2, 2])
}

# dgl.to_heterogenous using above graph
{
    '_ID': {
        'C': tensor([0, 1, 2, 3, 4]),
        'N': tensor([5]),
        'O': tensor([6, 7])
    }
}
```

The new MCTS node I would expect looks like so,

```
{
    'C': tensor([10, 11, 12, 13, 14]),
    'N': tensor([15]),
    'O': tensor([16, 17])
}
```

which I was able to accomplish using the code block we are commenting on.

Here is a visual of this graph with labeled nodes if this helps:
test_MUTAG_explain

File: python/dgl/nn/pytorch/explain/subgraphx.py (hunk 2 above)
Comment (mufeili) [reply python/dgl/nn/pytorch/explain/subgraphx.py:771]: No problems. I just want to confirm if this is expected.

File: python/dgl/nn/pytorch/explain/subgraphx.py (hunk 5 above)
Comment (mufeili) [reply python/dgl/nn/pytorch/explain/subgraphx.py:605]: I see. I think you are right. Thanks.

File: python/dgl/nn/pytorch/explain/subgraphx.py [hunk 6]
```diff
@@ -364,3 +365,445 @@ def explain_graph(self, graph, feat, target_class, **kwargs):
...
+                            ]
+                        ]
+                    ]
+                    for ntype, indicies in largest_cc_hetero.ndata[NID].items()
+                }
+            else:
+                available_ntypes = [
+                    ntype
+                    for ntype in new_subg.ntypes
+                    if new_subg.nodes(ntype).nelement() > 0
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/subgraphx.py:617]:
```suggestion
                    if new_subg.num_nodes(ntype) > 0
```

File: python/dgl/nn/pytorch/explain/subgraphx.py [hunk 7]
```diff
@@ -364,3 +365,445 @@ def explain_graph(self, graph, feat, target_class, **kwargs):
...
+                        new_subg.ndata[NID][ntype][
+                            new_subg_homo.ndata[NID][
+                                largest_cc_homo.ndata[NID][indicies]
+                            ]
+                        ]
+                    ]
+                    for ntype, indicies in largest_cc_hetero.ndata[NID].items()
+                }
+            else:
+                available_ntypes = [
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/subgraphx.py:614]: I assume `nx.weakly_connected_components` returns an empty list for a graph without edges?

File: python/dgl/nn/pytorch/explain/subgraphx.py [hunk 8]
```diff
@@ -364,3 +365,445 @@ def explain_graph(self, graph, feat, target_class, **kwargs):
...
+                    ntype
+                    for ntype in new_subg.ntypes
+                    if new_subg.nodes(ntype).nelement() > 0
+                ]
+                chosen_ntype = np.random.choice(available_ntypes)
+                # backtrack from subgraph node ids to entire graph
+                chosen_node = subg.ndata[NID][chosen_ntype][
+                    np.random.choice(new_subg.nodes(chosen_ntype))
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/subgraphx.py:622]:
```suggestion
                    np.random.choice(new_subg.nodes[chosen_ntype].data[NID])
```

File: python/dgl/nn/pytorch/explain/subgraphx.py [hunk 9]
```diff
@@ -364,3 +365,445 @@ def explain_graph(self, graph, feat, target_class, **kwargs):
...
+                ]
+                cc_nodes = {
+                    chosen_ntype: torch.tensor(
+                        [chosen_node],
+                        device=subg.ndata[NID][chosen_ntype].device,
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/subgraphx.py:627]:
```suggestion
                        device=subg.device,
```

File: python/dgl/nn/pytorch/explain/subgraphx.py [hunk 10]
```diff
@@ -364,3 +365,445 @@ def explain_graph(self, graph, feat, target_class, **kwargs):
...
+                )
+                _, out_neighbors = self.graph.out_edges(
+                    local_regions[src_ntype], etype=c_etype
+                )
+                local_regions[src_ntype] = list(
+                    set(local_regions[src_ntype] + in_neighbors.tolist())
+                )
+                local_regions[dst_ntype] = list(
+                    set(local_regions[dst_ntype] + out_neighbors.tolist())
+                )
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/subgraphx.py:479]: I think L474-479 are problematic, as you can mix the node IDs of different node types together.

**Comment by mufeili** [comment]
A high level comment: it might be easier to use homogeneous graphs as much as possible except for model computation.

File: python/dgl/nn/pytorch/explain/subgraphx.py (hunk 7 above)
Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/subgraphx.py:614]: `nx.weakly_connected_components` is actually able to handle graphs without edges (returns sets containing single nodes), the branch here was due to an error in `dgl.to_homogeneous` when trying to convert from a heterogeneous graph with no edges

File: python/dgl/nn/pytorch/explain/subgraphx.py (hunk 6 above)
Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/subgraphx.py]: done 👍

File: python/dgl/nn/pytorch/explain/subgraphx.py (hunk 9 above)
Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/subgraphx.py]: done, thanks!

File: python/dgl/nn/pytorch/explain/subgraphx.py (hunk 10 above)
Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/subgraphx.py:479]: what might an example of conditions for this? I figured mixing would not be possible when the type of `in_neighbors` are guarenteed to be `src_ntype` due to the `etype` scope when calling `graph.in_edges`? (and a similar argument for `out_neighbors`)

Comment (mufeili) [reply python/dgl/nn/pytorch/explain/subgraphx.py:479]: I see. You are right. Thanks.

File: python/dgl/nn/pytorch/explain/subgraphx.py [hunk 11]
```diff
@@ -364,3 +365,445 @@ def explain_graph(self, graph, feat, target_class, **kwargs):
...
+
+        split_point = num_nodes
+        coalition_space = {
+            ntype: list(
+                set(local_regions[ntype]) - set(subgraph_nodes[ntype].tolist())
+            )
+            + [num_nodes]
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/subgraphx.py:486]:
```suggestion
            + [split_point]
```
for readability

File: tests/python/pytorch/nn/test_nn.py [hunk 12]
```diff
@@ -1631,6 +1631,70 @@ def forward(self, g, h):
...
+        def __init__(self, in_dim, n_classes, canonical_etypes):
+            super(Model, self).__init__()
+            self.etype_weights = th.nn.ModuleDict(
+                {
+                    "_".join(c_etype): th.nn.Linear(in_dim, n_classes)
+                    for c_etype in canonical_etypes
+                }
+            )
+
+        def forward(self, graph, feat, eweight=None):
```
Comment (mufeili) [inline tests/python/pytorch/nn/test_nn.py:1664]: no need for `eweight`

Comment (ndbaker1) [reply tests/python/pytorch/nn/test_nn.py]: done 👍

**Comment by ndbaker1** [comment]
> A high level comment: it might be easier to use homogeneous graphs as much as possible except for model computation.

Ah ok I see how this could be easier/cleaner. I will keep this in mind going forward.  
Do you recommend refactoring the code?

**Comment by mufeili** [comment]
> > A high level comment: it might be easier to use homogeneous graphs as much as possible except for model computation.
> 
> Ah ok I see how this could be easier/cleaner. I will keep this in mind going forward. Do you recommend refactoring the code?

I'm fine either way. We can also do a second PR after we merge this.

**Review by mufeili — APPROVED** [review]
Thanks for the great efforts.

**Comment by mufeili** [comment]
@dgl-bot

**Comment by mufeili** [comment]
@dgl-bot

**Pull request merged**